From aa8bcbda2e6dc7d50a4ee5a5dd91a97ee4e11f6d Mon Sep 17 00:00:00 2001
From: thampan <joby.thampan@intel.com>
Date: Fri, 8 Nov 2019 11:10:32 +0800
Subject: [PATCH] DRVLIB_SW-3170: Fix for EPON Reconnect Issue

    - WAN Port need to have QoS Wred Port Config as 512 compared to other LAN port
      to avoid PON IP reconnect issue
    - Fixed the bug of ignoring Q31 for Setting WRED Queue Config
---
 .../net/ethernet/lantiq/switch-api/gsw_defconf.c   | 27 ++++++++++++++++++----
 1 file changed, 23 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c b/drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c
index 686fe45f30b5..6e0a6f6c23f2 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c
@@ -236,7 +236,10 @@ int gsw_misc_config(struct core_ops *ops)
 	return 0;
 }
 
-/* Default Qos WRED Config in switch */
+/* Default Qos WRED Config in switch
+ * API for setting the default QoS Wred Queue and Port Configuration
+ * for all 5 MAC ports and 32 Queues for Prx300
+ */
 int gsw_qos_def_config(struct core_ops *ops)
 {
 	GSW_QoS_WRED_PortCfg_t sVar;
@@ -247,13 +250,29 @@ int gsw_qos_def_config(struct core_ops *ops)
 		memset(&sVar, 0x00, sizeof(sVar));
 
 		sVar.nPortId = j;
-		sVar.nGreen_Min = 240;
-		sVar.nGreen_Max = 256;
+
+		/* For EPON Upstream Case with > 85% traffic,
+		 * GSWIP is having ACM Drop, which cause PON IP to get reconnected
+		 * To avoid this QOD WRED Port Cfg values for WAN Port
+		 * need to be increased to 512
+		 */
+		if (sVar.nPortId == MAC_2) {
+			sVar.nRed_Min = 500;
+			sVar.nRed_Max = 512;
+			sVar.nYellow_Min = 500;
+			sVar.nYellow_Max = 512;
+			sVar.nGreen_Min = 500;
+			sVar.nGreen_Max = 512;
+		} else {
+			sVar.nGreen_Min = 240;
+			sVar.nGreen_Max = 256;
+		}
 
 		ops->gsw_qos_ops.QoS_WredPortCfgSet(ops, &sVar);
 	}
 
-	for (j = 0; j < 31; j++) {
+	/* Setting Q WRED default config for Q0 to Q31 */
+	for (j = 0; j <= 31; j++) {
 		memset(&qcfg, 0x00, sizeof(qcfg));
 
 		qcfg.nQueueId = j;
